src/hardware/sblaster.cpp

```diff
@@ -2894,28 +2849,84 @@ bool SB_GetAddress(uint16_t &sbaddr, uint8_t &sbirq, uint8_t &sbdma)
 	return (sbaddr != 0 && sbirq != 0 && sbdma != 0);
 }
 
-static void sblaster_callback(const uint32_t length)
+static void sblaster_mixer_callback([[maybe_unused]]const int requested_frames)
+{
+	// We can ignore requested frames as this function gets called in a loop until it gets what it needs
+	// Overflow is not a concern as extra frames will remain in the channel's buffer and get mixed on the next callback
+
+	// We must not block in fast-forward mode or the timings used by the mixer get confused
+	if (MIXER_InChipmunkMode() && soundblaster_mixer_queue.IsEmpty()) {
+		sb.chan->AddSilence();
+		return;
+	}
+	const auto frames = soundblaster_mixer_queue.Dequeue();
+	if (!frames) {
+		// Queue must be stopped, otherwise Dequeue() will block until some frames are available
+		sb.chan->AddSilence();
+		return;
+	}
+	// Weird double de-referencing syntax as the object type is std::optional<std::unique_ptr<AudioVector>>
+	(*frames)->AddSamples(sb.chan.get());
+}
+
+static void generate_frames(const int frames_requested)
 {
-	auto len = length;
+	static int ticks_of_silence = 0;
 
 	switch (sb.mode) {
 	case DspMode::None:
 	case DspMode::DmaPause:
-	case DspMode::DmaMasked: sb.chan->AddSilence(); break;
+	case DspMode::DmaMasked:
+	// Enqueue a tick's worth of silenced frames
+	// Some games (Tyrian for example) will switch to DmaMasked briefly (less than 5 ticks usually)
+	// We can't use AddSilence on the mixer thread as that asks for a blocksize of audio (usually over 10ms)
+	enqueue_frames(std::make_unique<AudioVectorM8S>(
+		frames_requested,
+		std::vector<int8_t>(frames_requested).data()
+	));
+
+	++ticks_of_silence;
+	if (ticks_of_silence > 5000) {
+		// Most games using DMA will run it constantly in that mode
+		// If we've hit 5 seconds of silence, we're probably not playing a DMA mode game
+		// Stop the mixer channel for performance reasons as this channel blocks waiting for the main thread to provide more audio
+		// We want this to be a conservative value as disabling the channel clears any pending audio
```

Comment:
As per our chat earlier, many games use DMA to play digitsed sounds sporadically, while OPL or MIDI music is playing all the time. E.g., in Elvira, there are some speech samples during the intro and in-game you only get DMA-based sound effects during the battle scenes and on some special events.

But once a battle is over in 20-30 seconds, you'll usually spend the next few minutes to walking around, so only OPL music is playing. My point is, this DMA switching is a common occurrence in these games, so we need to make sure we don't drop audio when this happens (referring to the "clears any pending audio" part).

Reply:
> If we've hit 5 seconds of silence, we're probably not playing a DMA mode game

Maybe also make it clear that this enabling/disabling of the DMA channel will happen dynamically a lot in many games, so it's not only a startup thing as this comment implies.

Reply:
It checks for 5 seconds of silence so the only thing "pending" in that case would be silence.  That's what I meant by a conservative value.  A value of 5 seconds shouldn't ever cause a stutter.  I'll edit the comment though.

Reply:
Edited comment